pydarn/pydmap/datastructures.py

```diff
@@ -0,0 +1,212 @@
+# Copyright 2018 SuperDARN Canada
+# Author(s): Marina Schmidt
+"""
+This Module contains dmap data structure for pydarn.
+
+Classes
+-------
+DmapScalar : named tuple for DMAP scalar data properties
+DmapArray : named tuple for DMAP array data properties
+DmapRecord : Ordered dictionary that stores Dmap scalars/arrays by their name
+as the key and the data structure as the value.
+
+Exceptions
+----------
+TypeError
+    If something does not match the type setting
+
+See Also
+--------
+namedtuple
+OrderedDict
+typing : for type setting in python3
+
+Future Work
+-----------
+Relocation of module
+        If cdmap is implemented this may need to live in high folder level for
+        shared module and better organization.
+Review DmapRecord and its structure
+        Not sure if this structure is superfluous
+
+See DEVELOPER_README.md on more information regarding the data structures
+and their organization pertaining to DMAP records.
+"""
+from typing import NamedTuple, Type
+from collections import OrderedDict, namedtuple
+import numpy as np
+
+"""
+NamedTuple class that defines the structure of a DMAP scalar.
+All fields can be accessed using the dot syntax. See DEVELOPER_README
+for more information on NamedTuples.
+
+Fields
+-------
+name : str
+    the name of the scalar variable
+value : data type
+    the measure data value for the scalar
+data_type : int
+    the DMAP numerical value for the data_type
+"""
+class DmapScalar(NamedTuple):
+    name: str
+    value: Type[object]
+    data_type: str
+
+"""
+NamedTuple class that defines the structure of a DMAP Array.
+All fields can be accessed using the dot syntax. See DEVELOPER_README
+for more information on NamedTuple.
+
+Fields
+-------
+name : str
+    the name of the scalar variable
+value : data type
+    the measure data value for the scalar
+data_type : int
+    the DMAP numerical value for the data_type
+dimension : int
+    The number of dimensions of the array, e.i., 1D, 2D, 3D
+shape : list
+    A list of dimension sizes of the array
+"""
+class DmapArray(NamedTuple):
+    name: str
+    value: np.ndarray
```

Comment:
First off I really would like to see a DMAP example where  DMAP records are stored in records. The Spec may be outdated or have legacy code that I would not like to introduce into newer code. Especially if the legacy introduces potential bugs. 

I am more than happy to go back if a given example is present in the current DMAP format.

With that said, `ndarrays` can hold any type of object including strings and DmapRecords so there is no actual issue here (The bittersweet of python :p ). It just informs developers that the object field should be a data type of ndarray. Technically if they store a list or something else nothing will really happen because it is python. Just later on in the downstream something might throw a `TypeError` but that is the pythonic way.  

I think in later python versions type hinting might actually enforce type checking. 

https://docs.python.org/3/library/typing.html